e2e-tests/src/main.rs

```diff
@@ -0,0 +1,250 @@
+        balance_before,
+        balance_after,
+        transfer_value
+    );
+
+    Ok(())
+}
+
+#[derive(Debug, Decode)]
+struct NewSessionEvent {
+    session_index: u32,
+}
+
+fn test_change_validators(config: Config) -> anyhow::Result<()> {
+    let Config { node, seeds, sudo } = config;
+
+    let accounts = accounts(seeds);
+
+    let sudo = match sudo {
+        Some(seed) => keypair_from_string(seed),
+        None => accounts.get(0).expect("whoops").to_owned(),
+    };
+
+    let connection = create_connection(node).set_signer(sudo);
+
+    let validators_before: Vec<AccountId> = connection
+        .get_storage_value("Session", "Validators", None)?
+        .unwrap();
+
+    info!("[+] Validators before tx: {:#?}", validators_before);
+
+    let new_validators: Vec<AccountId> = accounts
+        .into_iter()
+        .map(|pair| pair.public().into())
+        .chain(iter::once(
+            AccountId::from_ss58check("5EHkv1FCd4jeQmVrbYhrETL1EAr8NJxNbukDRT4FaYWbjW8f").unwrap(),
+        ))
+        .collect();
+
+    info!("[+] New validators {:#?}", new_validators);
+
+    // wait beyond session 1
+    let current_session_index = wait_for_session(connection.clone(), 1)?;
+    let session_for_change = current_session_index + 2;
+    info!("[+] Current session index {:?}", current_session_index);
+
+    let call = compose_call!(
+        connection.metadata,
+        "Aleph",
+        "change_validators",
+        new_validators.clone(),
+        session_for_change
+    );
+
+    let tx = compose_extrinsic!(connection, "Sudo", "sudo_unchecked_weight", call, 0_u64);
+
+    // send and watch extrinsic until finalized
+    let tx_hash = connection
+        .send_extrinsic(tx.hex_encode(), XtStatus::Finalized)
+        .expect("Could not send extrinsc")
+        .expect("Could not get tx hash");
+
+    info!("[+] change_validators transaction hash: {}", tx_hash);
+
+    // wait for the change to be aplied
+    wait_for_session(connection.clone(), session_for_change)?;
+
+    let validators_after: Vec<AccountId> = connection
+        .get_storage_value("Session", "Validators", None)?
+        .unwrap();
+
+    info!("[+] Validators after tx: {:#?}", validators_after);
+
+    assert!(new_validators.eq(&validators_after));
+
+    Ok(())
+}
+
+/// blocking wait, if ongoing session index is >= new_session_index returns the current
+fn wait_for_session(
+    connection: Api<sr25519::Pair, WsRpcClient>,
+    new_session_index: u32,
+) -> anyhow::Result<u32> {
+    let module = "Session";
+    let variant = "NewSession";
+    info!("[+] Creating event subscription {}/{}", module, variant);
+    let (events_in, events_out) = channel();
+    connection.subscribe_events(events_in)?;
+
+    let event_decoder = EventsDecoder::try_from(connection.metadata)?;
+
+    loop {
+        let event_str = events_out.recv().unwrap();
+        let _events = event_decoder.decode_events(&mut Vec::from_hex(event_str)?.as_slice());
+
+        match _events {
```

Comment:
Isn't `if let` more idiomatic for `Result`?

Reply:
I'm also interested in the error variant - if it fails I'd like to see the reason logged

Reply:
Oh yes, my bad :)